### DISCUSSION TITLE
Auto pairing \(

### DISCUSSION THREAD

**Discussion by tarsobcaldas** [LunarWatcher/auto-pairs#52] [Q&A] [answered]
I write with LaTeX and I'm coming from Sublime Text. I used to auto pair \( while mapped from $. I'd like to achieve the same thing using vim, and found the original plugin. The problem was that it would't auto pair escaped characters. While reading [this issue](https://github.com/jiangmiao/auto-pairs/issues/325), I found this fork, with the possibility to write 
```
let g:AutoPairs = autopairs#AutoPairsDefine([{'open': '\\[', 'close': '\]', 'filetype': 'tex'}])
```
to auto pair '\[', but when I tried to substitute with \\(, I got the message 
```
Error detected while processing function autopairs#AutoPairsInsert:
line   15:
E55: Unmatched \)
```

Can someone help me with that? Also, what's the best way to map $ to \( while auto pairing?

**Comment by LunarWatcher** [comment]
Well, this is monumentally annoying. `\)` is unmatched and `\\)` inserts the string literal `\\)`. Smells like a bug, gonna need to dig into the code later to nail this one down.

As for the problem... this fork, still being based on large core parts of jiangmiao's code means it inherited large parts of the problems with jiangmiao, including backslashes. I'm not sure what regex group nightmare is going on here, but it's probably down to the matching regex somehow being badly interpreted. Worst-case, `\)` is interpreted as a regex group being closed in non-magic mode (which is a vim regex thing you don't really need to worry about). The TL;DR: is that it most likely tries to define the closure of a regex group, but without an opening, hence the "unmatched" part. And to add insult to injury, the _raw string itself_ is used in output, meaning that I probably need to add some escaping, or actually start working on some 4.0.0 logic, because backslashes are getting really annoying. 

---

Looked into it, and it was indeed regex: https://github.com/LunarWatcher/auto-pairs/blob/4dceac438bfd637db998b7b565e70bbe3e3142d9/autoload/autopairs.vim#L160

`\V` reverts to Vim's regular syntax, and guess what `\)` means in non-magic mode? 

I can make some temporary hacks here, but I'm suspecting the real fix is gonna be a part of 4.0.0, which is far back on my backlog for real-life reasons. Jiangmiao's decision to mix regex with non-regex keeps coming back to annoy the shit out of me. I'm genuinely inclined to make another breaking change and pull off a [rule-like system](https://github.com/cohama/lexima.vim#customization) instead. (Not along the lines where backspace and newline rules have to be explicitly defined though, because that just sounds like it requires a ton of unnecessary and verbose config).

Probably gonna settle for a hotfix on master for the time being, and actually deal with fixing the root problem later

**Comment by LunarWatcher** [comment, marked as answer]
Further bugs prevented this from working. As part of #71, this has been fully fixed with `'\\[': '\]'`. Changes to 4.0.0 (currently only part of the develop-4.0.0 branch) means the pair `'\[': '\]'` (and equivalent for `\(\)`) is valid, and the double backslash open is invalid as of 4.0.0. The changes allowing this have since been merged into main
